### PR TITLE
GPSampler as Default Optuna Sampler

### LINKED ISSUE #46: Make GP Sampler the standard Optuna Sampler when torch is installed
# Use `GPSampler` as the Default Optuna Sampler When `torch` Is Available

## Summary

MotherML currently hard-codes `TPESampler` as the default Optuna sampler in `MotherTuner`. The official Optuna documentation and benchmarks show that `GPSampler` is the better choice for budgets **up to ~500 trials**, which covers virtually all practical MotherML tuning runs (the default `n_trials_optuna` is 100). We should switch the default to `GPSampler` whenever `torch` (and `scipy`) is available, with a graceful fallback to `TPESampler` when it is not.

---

### Sampler Comparison Table

The [official Optuna sampler reference](https://optuna.readthedocs.io/en/stable/reference/samplers/index.html) contains an explicit **recommended budget** row:

| Sampler | Recommended Budget (# trials) | Time Complexity (per trial) |
|---|---|---|
| `TPESampler` | **100 – 1 000** | $O(d \cdot n \log n)$ |
| `GPSampler` | **up to ~500** | $O(n^3)$ |
| `CmaEsSampler` | 1 000 – 10 000 | $O(d^3)$ |

Where $d$ = search-space dimension, $n$ = number of completed trials.

**Key observation:** `GPSampler` is explicitly recommended for budgets ≤ 500 trials — the exact regime MotherML operates in by default. The $O(n^3)$ cost is negligible when $n < 200$ and the evaluation cost of a CatBoost/RF model dominates by orders of magnitude.

### Algorithm Selection Guide

The [Efficient Optimization Algorithms tutorial](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/003_efficient_optimization_algorithms.html) from the Optuna developers summarises:

> For limited parallel compute resources and no (or few) categorical hyperparameters:  
> **TPE. GP-EI if search space is low-dimensional and continuous.**

MotherML's CatBoost/RandomForest search spaces are exactly this: low-to-medium dimensional, mostly continuous/integer parameters (learning rate, depth, L2 reg, etc.), with only a handful of categorical choices (boosting type, grow policy). This is the ideal GPSampler regime.

---

## Evidence from AutoSampler (OptunaHub)

The Optuna team's own [AutoSampler](https://hub.optuna.org/samplers/auto_sampler/) — their recommended "just works" sampler — applies exactly this logic:

> **"Using `GPSampler` during the early stages of a search, as it has excellent sample efficiency but is only applicable to a small number of evaluations."**

Benchmark results from the [AutoSampler announcement blog post](https://medium.com/optuna/autosampler-automatic-selection-of-optimization-algorithms-in-optuna-1443875fd8f9) show:

- AutoSampler (GP-first) with **< 100 trials** finds a **better solution** than TPESampler with **1 000 trials** on multiple benchmark functions.
- AutoSampler converges significantly faster and is more stable (narrower variance bands) than default TPESampler on Hartmann6 and McCourt11 benchmarks.

This is the core argument: **GPSampler uses the trial budget dramatically more efficiently at small $n$**, because it builds a principled probabilistic surrogate (Matérn-5/2 GP with ARD) that models parameter correlations globally, whereas TPE's kernel density estimates only become reliable after ~50–100 observations.

---

## Why GPSampler Is Better in the Low-Budget Regime

| Property | TPESampler | GPSampler |
|---|---|---|
| Surrogate model | Kernel Density Estimates (l/g split) | Gaussian Process (Matérn-5/2, ARD) |
| Models parameter correlations | Partially (multivariate mode) | Yes, natively |
| Sample efficiency at $n < 100$ | Moderate | **High** |
| Sample efficiency at $n > 500$ | Good | Degrades ($O(n^3)$) |
| Acquisition function | EI via ratio of densities | log-EI (numerically superior) |
| Handles integers + floats well | Yes | **Yes** (better than CMA-ES) |
| Requires `torch` | No | **Yes** (`scipy` + `torch`) |

The GP's ability to reason about the global landscape from very few points is the fundamental advantage. TPE requires enough trials in each density bucket to estimate l(x) and g(x) reliably — this threshold is usually around 25–50 trials just for `n_startup_trials`, meaning effective Bayesian search only starts after that point.

---

## Fit for MotherML's Search Spaces

MotherML's CatBoost hyperparameter spaces (e.g., `learning_rate`, `depth`, `l2_leaf_reg`, `bagging_temperature`) are:

- Primarily **continuous and integer** — GPSampler's strength.
- **Low-to-medium dimensional** (typically 5–12 parameters) — well within the GP's reliable range.
- Not heavily conditional/tree-structured — GPSampler handles this adequately (vs. TPE's strength in deep conditional spaces).

The few categorical dimensions (`boosting_type`, `grow_policy`) are handled by GPSampler's independent fallback sampler (random by default), identical to how TPE handles them.

---

## Proposed Implementation

In `src/mother/optimization/core.py`, replace the hard-coded `TPESampler` default with a conditional:

```python
def _default_sampler(
    seed: int, n_startup_trials: int, multivariate: bool
) -> optuna.samplers.BaseSampler:
    """Return GPSampler when torch is available, else fall back to TPESampler."""
    try:
        import torch  # noqa: F401
        import scipy  # noqa: F401
        module_logger.debug("torch available — using GPSampler as default")
        return optuna.samplers.GPSampler(
            seed=seed,
            n_startup_trials=n_startup_trials,
            deterministic_objective=False,
        )
    except ImportError:
        module_logger.debug("torch not available — falling back to TPESampler")
        return optuna.samplers.TPESampler(
            multivariate=multivariate,
            group=True,
            constant_liar=True,
            seed=seed,
            n_startup_trials=n_startup_trials,
        )
```

Then in `MotherTuner.__init__`:

```python
if sampler is None:
    self.sampler = _default_sampler(
        seed=seed,
        n_startup_trials=n_startup_trials,
        multivariate=kwargs.get("multivariate", True),
    )
else:
    self.sampler = sampler
```

Users who need the old behaviour can still pass `sampler=optuna.samplers.TPESampler(...)` explicitly.

---

## Considerations / Open Questions

1. **`torch` is already a CatBoost dependency** in many environments; the incremental cost of this switch is zero for most MotherML users.
2. **GPSampler does not support pruning** (`NopPruner` is effectively required). MotherML does not currently use Optuna pruners, so this is a non-issue.
3. **GPSampler does not support batch/distributed optimization** well. MotherML runs `n_threads_optuna=1` by default; multi-threaded runs may want TPE with `constant_liar=True`. Could expose `sampler` selection as a config option tied to `n_threads_optuna`.
4. **`n_startup_trials`**: GPSampler defaults to 10 random warm-up trials; MotherML's current default is 12 — compatible.
5. **Categorical parameters** (boosting type, grow policy): GPSampler falls back to its `independent_sampler` (random) for categoricals. This is acceptable; TPE's handling of categoricals at small $n$ is not much better.

---

## References

- [Optuna Sampler Comparison Table](https://optuna.readthedocs.io/en/stable/reference/samplers/index.html)
- [Optuna — Which Sampler and Pruner Should be Used?](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/003_efficient_optimization_algorithms.html#which-sampler-and-pruner-should-be-used)
- [Optuna `GPSampler` API Reference](https://optuna.readthedocs.io/en/stable/reference/samplers/generated/optuna.samplers.GPSampler.html)
- [AutoSampler: Automatic Selection of Optimization Algorithms in Optuna](https://medium.com/optuna/autosampler-automatic-selection-of-optimization-algorithms-in-optuna-1443875fd8f9) (Ozaki, 2024)
- [Unexpected Improvements to Expected Improvement for Bayesian Optimization](https://arxiv.org/abs/2310.20708) — logEI acquisition used by GPSampler
- [Ozaki et al., Hyperparameter Optimization Methods: Overview and Characteristics, IEICE Trans, 2020](https://doi.org/10.14923/transinfj.2019JDR0003)

### PR DESCRIPTION
**Pull request by thomasATbayer**
# Branch: `46-make-gp-sampler-the-standard-optuna-sampler-when-torch-is-installed`

## Overview
This PR changes MotherML's default Optuna sampler selection when users do not pass a custom sampler.

- If torch is available: use `GPSampler`
- If torch is not available: use `TPESampler`
- If `GPSampler` initialization fails for any reason: fall back to `TPESampler`

The goal is to get stronger optimization in low-budget, expensive-trial settings, while keeping behavior robust across environments.

## Sampler decision logic (plain-English)
When `sampler=None`, MotherML now does the following:

1. If torch is available, try to build `GPSampler`.
2. If GP creation fails (dependency/runtime/init issue), fall back to `TPESampler`.
3. If torch is not available, use `TPESampler` directly.

This gives a safe default that prefers GP when possible but never hard-fails the run.

## Why this helps
MotherML tuning commonly has:

- relatively few trials
- non-trivial cost per trial
- mixed search spaces (shared parameters + conditional parameters)

`GPSampler` is often effective on the stable shared search space in this regime. But GP does not fully cover dynamic/conditional spaces. To avoid low-quality fallback behavior there, this PR configures a hybrid:

- GP for relative/shared space
- TPE as the independent sampler for startup and out-of-relative-space parameters

## Why GP + TPE together (and not GP alone)
`GPSampler` is strongest on a stable relative search space. In real MotherML tuning, many models (especially CatBoost-style spaces) include branching/conditional parameters.

For those parameters, GP does not always provide relative/joint suggestions. Optuna then delegates to the sampler's `independent_sampler`.

If that independent sampler is random, quality can drop on expensive low-budget runs. Using TPE here gives a history-informed fallback after startup instead of staying purely random.

## What changed in code
File: `src/mother/optimization/core.py`

1. Added explicit sampler builders:

- `_build_tpe_sampler(...)`
- `_build_gp_sampler(...)`

2. Default sampler selection logic in `MotherTuner.__init__` now:

- tries GP when torch is available
- catches GP initialization errors and falls back to TPE
- falls back to TPE when torch is unavailable

3. `GPSampler` is configured with:

- `independent_sampler=optuna.samplers.TPESampler(...)`
- the same `n_startup_trials` value as GP

4. Runtime safety for default path:

- GP init errors are caught and logged
- default path recovers to TPE rather than failing tuning setup

## Important behavior (reviewer-friendly)
This is the key runtime behavior for startup and conditional parameters:

1. Startup phase is controlled by global completed-trial count (study-level), not by per-parameter occurrence count.
2. During startup, independent sampling is used.
3. Because independent sampler is TPE, and TPE itself has a startup phase, TPE uses random sampling during its own startup window.
4. After startup, conditional/branch-only parameters are sampled by history-informed TPE instead of pure random.

### Detailed runtime flow for dynamic/conditional spaces
For each trial, sampling effectively works like this:

1. If the trial count is still in startup, independent sampling is used.
2. Once startup is over, GP proposes parameters in its relative/shared space.
3. Parameters not covered by GP's relative space (common in conditional branches) are sampled via the independent sampler.
4. Because that independent sampler is TPE, those uncovered parameters transition from random startup sampling to history-informed TPE sampling after startup.

Important reviewer note: startup cutoff is study-level trial count, not "how many times a specific conditional parameter appeared."

In short:

- Shared stable space: GP-optimized after startup
- Conditional/out-of-relative-space params: TPE, with random startup then history-informed behavior

## What this PR does not change
To keep scope focused, this PR is about default sampler strategy and safe fallback behavior.

It does not change broader optimization workflow semantics beyond sampler choice and GP/TPE wiring.

It also does not claim full joint GP modeling over all conditional branches. This is a pragmatic hybrid to improve defaults in mixed spaces.

## Test coverage
File: `test/unit/test_model_tuner.py`

Covered in `TestDefaultSamplerSelection`:

- GP selected when torch is available
- TPE selected when torch is unavailable
- GP init failure falls back to TPE
- GP dynamic-search-space warning behavior
- custom sampler passthrough

Also validated:

- callback behavior remains tested in `TestGetCallbacks`

## Trade-off summary
Pros:

- better default for small trial budgets on shared/stable spaces
- safer behavior for conditional spaces than GP + random fallback
- resilient fallback when GP stack is unavailable or fails

Caveat:

- this is a pragmatic hybrid, not full joint modeling of all conditional interactions by GP
- when large parts of the important search space are highly conditional/tree-structured, a fully TPE-centric strategy may still outperform GP-first defaults

## Suggested PR wording
This PR makes `GPSampler` the default Optuna sampler in MotherML when torch is available, with `TPESampler` as the independent sampler and as full fallback when GP cannot be initialized. The hybrid keeps GP benefits for low-budget optimization on stable shared spaces, while using history-informed TPE for startup and conditional parameters that fall outside GP's relative search space.

### CODE REVIEW

File: src/mother/optimization/core.py [hunk 1]
```diff
@@ -136,21 +136,35 @@ def __init__(
...
+                )
+            else:
+                module_logger.debug("torch not available — falling back to TPESampler")
+                self.sampler = optuna.samplers.TPESampler(
+                    multivariate=True,
+                    group=True,
+                    constant_liar=True,
+                    seed=seed,
+                    n_startup_trials=n_startup_trials,
+                )
```
[the remark this answers is not in the document]
Comment (thomasATbayer) [reply src/mother/optimization/core.py:199]: just increased the minimum rquirerd optuna version so this is covered now

[the remark this answers is not in the document]
Comment (thomasATbayer) [reply src/mother/optimization/core.py:199]: Scipy is already a Mother dependecny and also a dependency of quite a few other packages.

File: src/mother/optimization/core.py [hunk 2]
```diff
@@ -143,15 +173,30 @@ def __init__(
...
-                multivariate=kwargs.get("multivariate", True),
-                group=True,
-                constant_liar=True,
-                seed=seed,
-                n_startup_trials=n_startup_trials,
-            )
+            if torch_available:
+                try:
+                    module_logger.debug("torch available — using GPSampler as default")
+                    self.sampler = self._build_gp_sampler(
```
[the remark this answers is not in the document]
Comment (thomasATbayer) [reply src/mother/optimization/core.py:181]: This is now fixed in another pull request
